Fixing problem with too many open files #6642

mwiewior opened the pull request:
Fix #6642

droazen commented on src/main/java/org/broadinstitute/hellbender/engine/ReferenceContext.java:
```diff
@@ -425,4 +425,8 @@ public String getKmerAround(final int center, final int numBasesOnEachSide){
+
+    public void close (){
+        dataSource.close();
+    }
```
These context objects are not meant to be closeable. Instead, the engine takes care of closing the underlying datasource at tool shutdown.

Does this patch actually resolve your "too many open files" error in practice?

mwiewior replied:
It helped since the processing of my BAM file that locally (block size 32MB) consists > 500 partitions is now failing after processing ~400 partitions instead of ~30. But it still does not fix it. That was my first attempt. But the problem I think is that with ReferenceFileSource that it leaves open file descriptors and I guess the problem is only replicable if you process larger files locally. Not sure if this might be also related but ReferenceFileSource implements AutoClosable but it isn't surrounded with try-with-resources block: https://github.com/broadinstitute/gatk/blob/c918f8522f2bd5103509b7c04d4b2f7d3fa3eb1e/src/main/java/org/broadinstitute/hellbender/engine/spark/LocusWalkerSpark.java#L113 I tried to add it but then it start failing from the very beginning of the processing.

mwiewior replied:
@droazen could you please give me any hint so I can try to dig deeper to find the fix?

mwiewior replied:
@droazen could you pls somehow help me so that I could work on the fix ?

droazen replied:
@mwiewior We are still looking into this one, but I believe at this point @lbergelson has some ideas as to what might be going on. @lbergelson would you mind posting your thoughts on Spark file handle leakage here?

mwiewior replied:
Ok, thanks Guys @lbergelson @droazen  - if there is anything I can help with just ping me !

lbergelson replied:
@mwiewior I see what you're aiming at with this PR.  Each shard is creating a new reader that isn't getting closed.  I don't think making the context closeable makes sense though because in most of the codebase we would never want to close that (we typicall re-use the underlying datasource many times).  I think we need to restructure the map functions so the underlying reader is closed after the operations are complete instead of making the context closeable.

mwiewior replied:
@lbergelson @droazen  I think I found the root cause of the problem - I tried to  rewrite the code into combination of mapPartitions and flatMap so that one new reader is created per partition not per each shard? Could please take a look at it ? It solved the issue for me.

ldgauthier commented:
@mwiewior are you still interested in getting this merged?

mwiewior commented:
@ldgauthier I would love to !

mwiewior commented:
@ldgauthier I changed the PR as you can see and rewrote the method to minimize the number of open readers - could you please help me to get it finalized and merged?

mwiewior commented:
@ldgauthier could you please help me with that ? Is there anything I should change?

ldgauthier commented:
@mwiewior I haven't done any Spark at all, so I think that @droazen and @lbergelson would be better reviewers.  I did see that your conda and CNN tests appear to be failing, though you didn't change anything that should affect Python.  I gave them a kick because hopefully it was just an inconsequential random error.

mwiewior commented:
@droazen @lbergelson so may I ask for reviewing it again ?
